### ISSUE TITLE
Make modifiers api simpler & uniform

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx#211]
I think the use of modifiers can be made more intuitive. In short, instead of the (imho) weird wrapping functions as modifiers, use a settings object that can be passed to both `extendObservable` and `(@)observable`. 

Some examples:

(ES6) Create an observable reference to a non-observable array:
`@observable prop = asReference([])` -> `@observable({ reference: true }) = []`

(ES6) Create an observable array, but do not make the children observable:
`@observable prop = asFlat([])` -> `@observable({ shallow: true }) = []`

(ES5) Create an observable reference to a non-observable array:
`extendObservable(this, { prop: [] }, { reference: true })`

This means that you will sometimes need two `extendObservable` calls where previously one would do, but I think the api will be cleaner.

So the modifiers would get a slight rename as well: `asStructure` ->  `struct`, `asReference` -> `reference`, `flat` -> `shallow`.

It could be interesting to merge the concepts of `shallow` and `reference`, as I don't think anybody is using shallow objects, and for primitive values shallow and reference behave the same. But in that case it would not be possible anymore to create references to non-observable arrays anymore.

Issues #197 #173 #55 should be addressed by these changes.

What do you guys think? Suggestions are welcome!

**Comment by AriaFallah** [comment]
I'm not so sure about this one. For example in my code for mobx-store I'm using

```js
db.set(key, createData(db, key, value.__data))
```

where createData is

```js
export function createData(obj: Object, key: string, value: Array<any> = []) {
  return {
    obs: value, // The observable array
    __data: asReference(value) // The underlying array
  }
}
```

If you change the API, how would you handle cases like this?

Also in the case that you have something like

```js
observable({
  a: asFlat([]),
  b: asStructure({}),
  c: asReference([])
})
```

I'd need to call extendObservable 3 times, which doesn't seem so intuitive.

As a side note:

The names `asStructure` and `struct` never made that much sense to me. Something with `deep` in it is a bit more intuitive considering it's deeply comparing the observable value to the previous one.

**Comment by hnordt** [comment]
I liked te idea. The current modifiers API is weird, the new proposal are cleaner.

**Comment by AriaFallah** [comment]
@mweststrate Would it be possible to just have both? Keeping them as exports, but also updating the `@observable` decorator?

**Comment by mweststrate** [comment]
@AriaFallah it will stay at least in the 2.0 range. If it should stay longer then that, maybe the `(extend)Observable` functions shouldn't be able to accept these options as arguments, to don't have confusingly two ways of achieving the same thing.

**Comment by mdebbar** [comment]
My 2 cents.

Would it be possible to do away with the boolean settings, and instead expose hooks that developers can use to customize the observation process. That way `asMap`, `asReference`, and more can be implemented in user land, outside the core of mobx.

I'm not familiar with the internals of MobX, but this is roughly the idea:
```javascript

/** This code can be implemented as a 3rd party library **/
function observableMap(observed) {
  return {
    get: (key) => observed[key],
    set: (key, value) => {
      observed[key] = observable(value);
    }
  };
}
const asMap = { init: observableMap, onObserve: ... };
/** ------ **/

@observable(asMap) prop = {};
```

**Comment by urugator** [comment]
First of all, please stop advertising decorators as the ES6 way. Decorators are not part of the standard, not even in ES2016 (as documentation claims). No matter how awesome decorators are, I don't think they should be advertised as preferred way of doing things (in case of JS users). Same goes for field initializers.
`extendObservable` is ES6 and should be mentioned first in docs, not somewhere below in "utils" section.
Even though MobX looks nice and easy to use with decorators and field initializers, they makes it harder for JS users to start with and adopt MobX library. 

For the modifiers, I would like to present some opinions and than an idea for implementation:

**1)** Modifiers API should be **"additive"**, so that complexity is added with each modifier, not the other way around.
For example an array is currently converted into observable array by default. I need to "undo" or "turn off" functionality by modifiers to make it less complex (flat/reference).
I propose to change it - one modifier makes property observable, another adds complexity by converting the value to observable array and so on. Will it be annoying to apply multiple modifiers most of the time? You won't be applying multiple modifiers, you will apply one, which merges behavior of others. The difference is that modifier can only  **add** behavior, not remove.

**2)** API should be **explicit**.
We don't need to rely on single magic `observable` function with some default conversion rules overridden by modifiers. We can be explicit and make everything crystal clear, consider something like this:
```
@mobx.ref p = {}; // same as asReference   
@mobx.ref p = []; // same as asReference
@mobx.shallow p = {}; // same as asFlat     
@mobx.shallow p = []; // same as asFlat   
@mobx.deep p = {}; // same as no modifiers   
@mobx.deep p = []; // same as no modifiers
```
Notice that `@mobx.deep` can work the same way as current `@observable` without modifiers, but it is much easier to predict the behavior.

**3)** Modifiers should be **composable**.
Something like this should be possible `@mobx.shallow(mobx.struct)`.
This goes hand in hand with "additive" nature.  

**4)** Modifiers API should be **extendable**.
I should be able to easily define new modifiers and reuse the existing ones.  

### How to implement it:
Let's define an object which I call "**observable property descriptor**".
It is and object which describes how the observable property is being set, retrieved, compared and is internally used for doing so.  It might look like this:
```
{
  set(value): {...},
  get(): {...},
  equals(value, other): { ... }     
}
```
We would have some generic decorator `@mobx.prop` (or @mobx.observable) which generates the most basic **observable property descriptor**, representing observable property (prop is observed as reference, value is kept as is, type does not matter, no conversion rules). Let's call it **default descriptor**.
We also define **modifier** as a function which **takes one descriptor and returns another**:
`descriptor -> descriptor`
So let's implement a `mobx.shallow` modifier (for simplicity only array values are supported): 
```
mobx.shallow = function(descriptor) {   
  // we will override "set" so that it first wraps the value into (shallow) ObservableArray and than delegates the call to the original "set"
  return Object.assign({}, descriptor, {    
    set(value) {
      if (Array.isArray(value)) {        
        return descriptor.set(new ObservableArray(value)); // wrap and delegate
      } else {
        throw new Error("Not implemented yet");
      }      
    }
  });                 
}; 
```
Now we can use the modifier like this `@mobx.prop(mobx.shallow) prop = [];` What's happening here? 
As mentioned before, `@mobx.prop` creates the **default descriptor**, but instead of immediately using it, it invokes passed **modifier**, providing default descriptor as an argument `mobx.shallow(defaultDescriptor)`, and uses returned descriptor instead of the default one.
This is fine, but it's not composable. I would like to be able to do this `@mobx.prop(mobx.struct(mobx.shallow))`. That's a bit tricky.
I would either have to change the syntax, so that it's always required to invoke the modifier: `@mobx.prop(mobx.struct(mobx.shallow()))`. I woudn't say it looks worse, but I can imagine a user can easily forget the brackets.
That way, modifier implementation stays quite simple:
```
mobx.shallow = function(modifier) {
  return descriptor => {
    descriptor = modifier ? modifier(descriptor) : descriptor;
    return Object.assign({}, descriptor, { /* same as before, omitted for simplicity */ });
  };
}
```
Otherwise the modifier return value would have to change based on the passed argument: `modifier -> modifier` and `descriptor -> descriptor`
An implementation is a bit more complicated:
```
mobx.shallow(arg) {
  const modifier = function(descriptor) => {    
    return Object.assign({}, descriptor,  { /* same as before, omitted for simplicity */ });
  }   
  if (typeof arg === "function") {    
    return descriptor => modifier(arg(descriptor));    
  } else {
    return modifier(arg);
  }  
}
```
I prefer the first option, because I don't like the ambiguity of the second.
Maybe there is a better way, which I don't know about... 

Now, if I use `@mobx.prop(mobx.struct(mobx.shallow()))` often, I can combine the two into one modifier `@mobx.prop(shallowStruct())`:
```
const shallowStruct = (modifier) => mobx.struct(mobx.shallow(modifier)); 
```
Or in the end, it should be possible to implemenet an own decorator `@shallowStruct`.

Additionally it may be also possible to obtain modifier from passed value:
```
@mobx.prop prop = {
  /* 
    This might be some custom Map implementation
    The object itself defines how it wants to be handled, when assigned to observable prop
  */
  [mobx.Symbol.propertyDescriptor](descriptor) { 
    //...
    return descriptor;
  }  
}
``` 
Due to the composability, both modifiers can be applied, the one specified by value itself, and those passed to decorator.

Note that all above is just an idea, it might require singnificant changes or might not be possible at all, but I hope there are some interesting thoughts at least.

**Comment by mweststrate** [comment]
@urugator the reason for the current behavior of `observable` is convention-over-configuration, so it acts like a swiss knife, trying to apply the most commonly wished behavior to whatever you pass to it. This often works great, but when being more experienced with MobX it is also nice to make things explicit indeed.

I think technically it should be very doable what you want, and I like the idea of moving the modifiers to the decorators and do that in a composeable way. What I am missing in your proposal is a way to make this convenient for ES5 / ES6 users as well. Any ideas about that?

Modifiers work ok-ish now, but they are not nice from both the perspective of strong typing and the fact that their behavior is sticky (mostly) isn't nice either. (Stickyness is fine, but it is not clear from the syntax atm)

(P.s. if ES6 is still mentioned anywhere in the docs where it should be ESNext, please file a PR!)

**Comment by urugator** [comment]
I had some ideas and I like this one the most:
```javascript
this.prop1 = {};
this.prop2 = [];
mobx.makeObservable(this, {  
  prop1: mobx.shallow(mobx.struct()),
  prop2: mobx.deep(),      
  //propName: modifier    
})

```
It's similar to decorators in a way, that it just adds some meta informations to existing fields.
However someone might not like that the definition of a single field needs to be on two places.
That can be solved like this:
The observable property descriptor also contains a field `value`, which represents initial value of the field.
So all we have to do is to a create a function accepting a value and returning modifier which sets that value to descriptor:
```javascript
// Function returning modifier 
mobx.val = function(value) {
  return descriptor => {
    descriptor.value = value;
    return descriptor;
  };
};
// Usage
mobx.makeObservable(this, {  
  prop: mobx.shallow(mobx.val([])),  
})

```
We could make the syntax even shorter, by letting modifiers to also accept value: `mobx.shallow([]);`
However, because the value can be function and modifier is also a function, there is no way to differentiate between the two. So a change is required:
```javascript
// Modifier is now an object
class ShallowModifier extends Modifier {
  constructor(wrapped) {
    this.wrapped = wrapped;
  }
  modify(descriptor) {
    // Now we can check for type    
    if (mobx.isModifier(this.wrapped)) {
      // and either use wrapped modifier
      descriptor = wrapped.modify(descriptor);  
    } else {
      // or set initial value
      descriptor.value = this.wrapped 
    } 
    // Shallow implementation, same as before        
    return Object.assign({}, descriptor, {    
      set(value) {
        if (Array.isArray(value)) {        
          return descriptor.set(new ObservableArray(value));
        } else {
          throw new Error("Not implemented yet");
        }      
      }
    });    
  } 
}

// Actual modifier function become: 
mobx.shallow = function(arg) {
  return new ShallowModifier(arg);
}

// The mobx calls modifier.modify(defaultDescriptor) instead of modifier(defaultDescriptor)
```
This change also makes it easier to create modifiers because a lot of the logic can be moved to parent:
```javascript
class Modifier {
  constructor(wrapped) {
    this.wrapped = wrapped;
  }
  modify(descriptor) {
    // Now we can check the type
    if (mobx.isModifier(this.wrapped)) {
      // and either use wrapped modifier
      descriptor = wrapped.modify(descriptor);
    } else {
      // or set initial value
      descriptor.value = this.wrapped 
    }
    return descriptor;         
  } 
}

class ShallowModifier extends Modifier {
  constructor(wrapped) {
    super(wrapped)
  }
  modify(descriptor) {              
    return Object.assign({}, super.modify(descriptor), {    
      set(value) {
        if (Array.isArray(value)) {        
          return descriptor.set(new ObservableArray(value));
        } else {
          throw new Error("Not implemented yet");
        }      
      }
    });    
  } 
}
```

**Comment by mweststrate** [comment]
@urugator would you be interested in taking a shot at it? Branch https://github.com/mobxjs/mobx/tree/fix-571-no-instanceof is probably the most convenient one to start with. If you are unfamiliar with TS, feel free to leave the typings (cast (`as any` where needed)). If you would manage that would be great!

**Comment by mweststrate** [comment]
Idea in #572: support `name` modifier

**Comment by urugator** [comment]
#572 brings another requirement:  Modifiers should be  **parameterizable**.
Now we need to somehow deal with wrapped modifier, initial value and optional parameters.
Clearly, proposed syntax won't handle that. What are the options:

#### Passing complex "options" object:
```javascript
mobx.name({ name: "name", wrapped: mobx.shallow({ wrapped: mobx.struct({ wrapped: mobx.val(5) }) }) })
```
**Problem:** complex syntax, far from declarative

#### Modifier accepts parameters, wrapping is done explicitely via wrap method:
```javascript
mobx.name("name").wrap(mobx.shallow()).wrap(mobx.struct().wrap(mobx.val(5))
```
**Problem:** quite complex syntax, modifier must be mutable (don't like it) 

#### Parameterizable modifiers are created via factory:
```javascript
mobx.name("name")(mobx.shallow(mobx.struct(mobx.val(5)))
```
**Problem:** ambiguity in usage - `parametrizedModifier(param)(wrapped)` VS `nonParametrizedModifier(wrapped)`

#### All modifiers are created via factory:
```javascript
mobx.name("name")(mobx.shallow()(mobx.struct()(mobx.val()(5)))
```
**Problem:** brackets hell

#### Convention - last parameter is wrapped modifier
```javascript
mobx.name("name", mobx.shallow())
```
**Problem:** imo bad practise, prone to all sorts of problems   

### New proposal:
#### Modifiers can't be composed on their own, but are composed by something else:
```javascript
// Composition is done by decorator   
@mobx.prop(mobx.name("name"), mobx.shallow(), mobx.struct()) prop = 5;
// Composition is done by extendObservable
mobx.extendObservable(this, {  
  prop: [mobx.name("name"), mobx.shallow(), mobx.struct(), mobx.val(5)]
}); 
```
**Problem:** I don't see any, I think it's quite fine - syntax is clear, declarative, quite short, modifier itself is simpler.
However there is still a way to make syntax shorter by making modifiers **chainable**:
```javascript  
// decorator   
@mobx.prop(mobx.name("name").shallow().struct()) prop = 5;
// ES5
mobx.extendObservable(this, {  
  prop: mobx.name("name").shallow().struct().val(5)
}); 
```
**Problem:** custom modifiers can't be chained
Solutions:
a) don't chain them, use previous approach:   
```javascript
const modifier1 = require("CutomModifier1");
const modifier2 = require("CutomModifier2");
@mobx.prop(mobx.name("name").shallow().struct(), modifier1(), modifier2()) prop = 5;
```
b) use some helper function for extending default "factory"     
```javascript
const mobx = require("mobx");
// Create customized factory
const modifiers = mobx.modifiers({
  modifier1: require("CutomModifier1"),
  modifier2: require("CutomModifier2"),
});
@mobx.prop(modifiers.name("name").modifier1().struct().modifier2())
```
Most of the time user will probably use one to two modifers max, so chaining might be really unnecessary, however I would at least provide `val` for each modifier (via extendable Modifier) as a convinient way to set initial value inline for ES5 users: 
```javascript
// Instead of the need to use two modifiers to define initial value inline
mobx.extendObservable(this, {  
  prop: [mobx.name("name"), mobx.val(5)]
}); 
// you can use one 
mobx.extendObservable(this, {  
  prop: [mobx.name("name").val(5)]
});
```
(Chainable modifier is still just normal modifier with additional methods used for chaining, so it can be introduced anytime, by anyone, without the need to change anything in modifiers concept)

One important thing to note:
Currently `observable` and `@observable` are the same functions and the behavior is determined based on the number and type of the passed args.
I don't think it's very clear solution, but regardless of what I think, it's not possible to do the same with modifiers, if they are supposed to accept arbitrary parameters.
Therefore modifiers and decorators must be placed in separate namespaces and must be used as two different things (which they are).
For example it's not possible to use `shallow` as decorator on one line and the same `shallow` as modifier on another.

**Comment by mdebbar** [comment]
I'm not a big fan of solution b) to chaining. That creates a loose dependency between stores and modifiers and it could easily break. It makes it hard to follow where a modifier is defined. I very much like the idea of a "modifier factory" instead. I look at it as a higher-order function that takes a simple modifier function and wraps it with all the necessary logic to handle decorator vs. modifier, passing extra options, etc.

**Comment by urugator** [comment]
> loose dependency between stores and modifiers and it could easily break

Not sure what you mean, stores are always dependent on modifiers they use.
Option b) just shows a possible way of making the custom modifiers chainable together with built-in modifiers.
I am not a big fan of it either, but I don't think it introduces some new dependencies. To make it clear some pseudocode:
Normal unchainable modifier:
```javascript
class MyModifier extends Modifier {
  modify(descriptor) { /* modifier logic */ }
}
```   
Chainable modifier:
```javascript
class GenericChainableModifier extends Modifier {
  // this modifier accepts a set of modifiers it can use and chain (for example modifiers required by store)
  constructor(modifiers) {
    // Add methods for chaining 
    for (let modifierName in modifiers) {
      let clazz = modifiers[modifierName];      
      this[modifierName] = (...args) => {        
        this.usedModifiers.push(new clazz(...args));
      }
      return this;  
    }
    // Track which modifiers were chained
    this.usedModifiers = [];    
  }  
  modify(descriptor) {
    // Apply all used modifiers       
    for (let modifier of this.usedModifiers) {
      descriptor = modifier.modify(descriptor);
    }
    return descriptor;
  }  
}
```
So now you just need a method, to which you can pass required modifiers and it will create such chainable generic modifier, which is exatly what I tried to illustrate in option b)
```javascript
function modifiers(modifiers) {
  // Let's make built-in modifiers availbale by default, so that user doesn't have to require them  
  const modifiers = mobx.getBuiltInModifiers();
  // Add custom modifiers provided by user
  foreach (let modifierName in modifiers) {
    modifiers[modifierName] = modifiers[modifierName];
  }
  // Create chainable modifier with additional methods representing required modifiers  
  return new GenericChainableModifier(modifiers); 
}
```
Here is usage example with store:
```javascript
// Normally you could simply require the modifiers like this:
// const modifier1 = require("CustomModifier1");
// const modifier2 = require("CustomModifier2");
// However we want to make them chainable, so we will create generic chainable modifier using the method above
const modifiers = modifiers({
  modifier1 = require("CustomModifier1");
  modifier2 = require("CustomModifier2");
});
class MyStore {
  @prop(modifiers.modifier1().modifier2().builtInModifier()) prop = 5;
}
// As you can see, the dependencies are the same, they are just not referenced directly,
// but through "modifiers" object, which makes chaining possible
// Personally I would never use the first approach, because I don't like to pollute the local scope with a lot of functions
// so I would probably still access modifiers through some namespace object,
// therefore, from my point of view, the only difference is the chaining ability
```
Don't get distracted too much with the chaining, it's not a solution to modifiers itself, it was just an idea for shorter syntax and as you can see anyone can add this functionality himself.

> Simple modifier function

I am convinced that modifiers should be objects, not simple functions. It's more robust solution, providing type safety, own properties, ability to inherit functionallity from parent, etc...
You will be implementing own modifiers rarely, but it's still very easy, just extend class and overwrite `modify` function.

> "modifier factory"

Could you please provide an example of what you exactly mean. I am also interested in what you mean by "logic to handle decorator vs. modifier"

**Comment by mdebbar** [comment]
### Loose dependency
Never mind, I misunderstood. But I still think chaining doesn't add a lot value here.


### Function vs. object modifier

#### 1. Inheritance:

Anything you want to achieve with objects, you can achieve with functions (and composition). For example, if you want to extend/inherit the logic of another modifier:
```javascript
const existingModifier = require("...");

module.exports = function newModifier(descriptor) {
  var newDescriptor = existingModifier(descriptor);
  /* do more logic with the new descriptor */
  return newDescriptor;
};
```
#### 2. Modifier factory:
Using classes and objects makes it easy to put all the common logic in the parent class, and all modifiers inherit from there. But we can achieve the exact same thing with just functions:
```javascript
// Using classes:
class Modifier1 extends Modifier {
  modify(descriptor) { /* modifier logic */ }
}
// Using functions:
const modifier1 = createModifier(function(descriptor) { /* modifier logic */ });
```
Now the question becomes, how does `createModifier` work? The easiest implementation would be:
```javascript
function createModifier(modifierLogic) {
  return class MyModifier extends Modifier {
    modify(descriptor) {
      return modifierLogic(descriptor);
    }
  }
}
```
It's creating a class under the hood, but that's hidden from the user.

#### 3. Decorators vs. modifiers:
Now that we have a factory called `createModifier`, we can rely on it to do the heavy lifting of distinguishing between the decorator/modifier use cases:
```javascript
function createModifier(modifierLogic) {
  function modifier(wrapped) {
    return function() {
      if (decoratorArguments(arguments)) {
        // ... use `wrapped` and `modifierLogic` as decorator.
      } else {
        var descriptor = arguments[0];
        // ... use them as regular modifier.
      }
    };
  }
  return modifier;
}
```

**Comment by urugator** [comment]
> Anything you want to achieve with objects, you can achieve with functions (and composition). 

Functionally yes, however I wouldn't say it's always the most practical. 
   
> 2. Modifier factory:

It's good pattern to add functionality by using composition, I woudn't use it for this though (building class). I can eventually explain why, but for now, it's not so important.

Could you please change your examples in a way, that it's possible to pass any arbitrary arguments to modifier, for example argument `name`?
Also provide usage of that custom modifier together with another modifier for example `shallow()`.
And prefix functions from mobx realm with `mobx.` so that it's clear what's implemented by user and what's by MobX.

**Comment by mweststrate** [comment]
I'm not sure whether generic modifiers are really required as feature; their meaning has usually a lot of impact on the implementation of the actual types in mobx. (e.g. `asStructure` behaves very different for computed values and objects).

The most important thing is to have better names, and nice syntax for ES5 I think. 

Although, custom modifiers could be nice sugar around attaching `intercept` or `observe` hooks.  (that can do a lot but not change the internal semantics of mobx).

That being said, I didn't read the latest comment yet, so I might be just repeating others ;-)

_/me scrolls back to top_

**Comment by mweststrate** [comment]
not sure btw whether #572 should be solved via identifiers; setting it at declaration time or with a decorator might be nicer:
```javascript
var todo = observable("Todo", { title: bla }) // or a s second arg. same for extend observable
```

For ES.next it could be a class decorator maybe

**Comment by mweststrate** [comment]
Personally I would I think have the following contract for modifiers:
```
type IMobxPropertyDescriptor = {
  initialValue?: any
  shallow?: boolean
  name?: string
  compareStructural?: boolean
  interceptor: change => change
  observer: change => void
}

type IMobxModifier = (....customArgs: any[], IMobXPropertyDescriptor | /*value*/ any) => IMobXPropertyDescriptor
```

Usage:
```
@observable(shallow()) prop = 7
@observable(name("myname", shallow()) prop = 8

const x = observable(name("test", {
  field: shallow(7),
  otherfield: myCustomModifier(someValue)
})
```

Definition:
```
function shallow(thing) {
   const descriptor = isMobxPropertyDescriptor(thing) ? thing : { initialValue: thing }
   return Object.assign(descriptor, { shallow: false })
}

var myCustomModifor = createInterceptMofifier(myCoolFunction)

function createInterceptModifier(func) { // built-in utilitiy
   return function(value | propertyDescritpor) {
      // something similar to shallow
   }
}
``` 

have to go, so this is basically a brain fart ;-)

**Comment by urugator** [comment]
> not sure btw whether #572 should be solved via identifiers; setting it at declaration time or with a decorator might be nicer

Imo it's just a matter of time until another similar requirement will appear, therefore I woudn't solve it by introducing another parameters...
The problem can be generalized: "How to provide some metadata for observable object?". I would use the same concept as with properties - descriptor and modifiers.
Let `observable` accept modifier(s) the same way as `extendObservable` does: `observable(value, name("name"))`

> contract for modifiers:

I don't like it, because it operates on the defaults defined by `observable`, making the modifier behavior incomposable, basically only flag swithing is possible. 

So I still propose this contract:
```javascript
type IMobxPropertyDescriptor = {
  name?: string
  value?: any
  set?: value => void
  get?: void => value 
  equals?: value, other => boolean   
  observer?: change => void
  owner?: object // reference to object owning the property
  modifiers: array // array of already applied modifers
}
```
`interceptor` is not needed, because we've got `set` and `get`, they do the intercepting.
`compareStructural` is also not needed, because we can provide any comparator via `equals`.
`intialValue` is renamed to just `value`, because it may be overwritten by modifier (maybe one day, we might need need both initial/modified)
Added `owner` - reference to property owning object, useful for debugging, provides ability to access owner's props/methods 
Added `modifiers` - array of references to already applied modifiers, useful for debugging, warnings etc 

Pseudocode for descriptor generated by default, without any modifiers applied:
```javascript
{
  name: propertyName,
  value: propertyValue,
  get() {
     return this.$mobx.getPropertyValue(propertyName);
  }
  set(value) {
     this.$mobx.setPropertyValue(propertyName, v); // value is not made observable by default
  }
  equals(value, other) {
    return value === other;
  }
  owner: this,
  modifiers: []
}
```

`observable` and `@observable` additionally applies modifier which roughly looks like this: 
```javascript
// inside ObservableModifier class definition
modify(descriptor) {
  return Object.assign(descriptor, {
    set(value) {    
      if (this.asFlat) {
        value = asFlat(value);
      }
      if (!this.asReference) {
        value = observable(value); // make value itself observable
      }    
      descriptor.set(value);
    }
    equals(value, other) {    
      if (this.asStructure) {
        return compareAsStructure(value, other);
      }
      return descriptor.equals(value, other);
    } 
  }); 
}

// In reality you could simply reuse "deep" and "structure" modifiers
``` 
However I would get rid of `@observable` altogether and replace it with `@deep`, because that's exactly what it does.
Again, modifiers should add or overwrite behavior, but not to remove it - `shallow` should make value itself shallowly observable, not to turn off deep observability.

For the the usage:
```javascript
// property observable as reference
@prop prop = 9;
// shallowly observable property 
@shallow prop = 9;
// which is just a shortcut for
@prop(shallow()) = 9;
// property observable the same way as current @observable without modifiers
@deep prop = 9;
// which is just a shortcut for
@prop(deep()) prop = 9;
// multiple modifiers
@prop(name("name"), deep()) prop = 9;  
// same as
@deep(name("name")) prop = 9;

// the same with ES5
extendObservable(this, {
  prop: 9,
  prop: shallow(9),
  prop: deep(9),
  prop: [name("name"), deep(9)]
});
```
For clarification, modifier itself is an object implementing `modify` method and the functions above (`shallow()`,`deep()`, etc) instantiates these modifiers.

I am already repeating myself here, so I will try to not post another wall of text for some time. Sorry for these btw, it's hard to express thoughts textually and concisely :)

**Comment by mweststrate** [comment]
yeah no problem, I think you are onto something here, but I am sadly too occupied atm to take a deep dive into this :) (4 talks in the next 3 weeks), but I'll come back to this later! In november or something like that

**Comment by antitoxic** [comment]
I am throwing my thoughts here. I know they might not be very insightful. They are just a perspective.

I might have also misunderstood some of the ideas so far. Anyways, here's a few things:

I would prefer to keep `observable()` as my main entry point for making things observable.  Not `.prop` which is way too generic or `.ref`, `deep` which are different ways to define specialized version of `observable()`.

I mean, `observable` can be renamed to something else but decorating property or value with `prop()` doesn't make linguistical sense to me. When I read it, I don't immediately understand what it does where reading @observable is very clear.

I'm thinking this while having python zen in heart: we should preferably have one obvious way to do one thing. Not 5 ways that do *almost* the same thing. 

I would enjoy if `observable()` is configurable but instead of settings like `asStructure: true` or `struct: true` we can pass references to mobx or custom observable types. E.g. `observable(mobx.shallow, mymodule.haveId)` or `observable(mymodule.haveId(mobx.shallow))`.

Basically modifiers are just equality strategies. I like the `{ ... equals:()}` as part of a descriptor. I also like that it removes the `interceptor` concept. **Less new concepts, same functionality**. I like the idea of a descriptor as a whole because I had problems with mobx/babel/typescript not letting me change the descriptor on the property. 

Mobx is all about tracking change. I don't need it to be a generic decorator/transformation library. I need it to help me track and optimize changes across my app state. `@track`? :) 

Anway I say this all with a combo of frustration and ideas. Hope it makes sense.

**Comment by urugator** [comment]
> `observable` instead of `prop`

Initially I wanted to use just `observable`, but to avoid a confusion with current `observable`, I decided to use different nomenclature, which would implicate more clearly the suggested changes in implementation. 
Another concern was alignment with actual `observable` method and backward compatibility.

> linguistic sense 

The most descriptive counterparts would be something like `@observableProperty`, `@deeplyObservableProperty`, etc.
I quess we will agree that using term "property" on **property decorator** is redundant. 
However I feel like using term "observable" is kind of redundant too.
The thing is, the linguistic sense is actually hidden in the import statement.
You know, you're using a module for handling observability and the observability is the only responsibility of that module.
So if you use `mobx.prop`, the `mobx` part more or less already contains that "observable" word.
Consider class `Observable` with a static method `from`. If you `import { from } from "Observable"`, then using only `from` won't make sense either.
At the same time, it doesn't make sense to add static method `observable` just so you could use `import { observable } from Observable`.
Conclusion: Yes, using these terms alone doesn't make sense, the intended real life usage is within a namespace: `@mobx.prop` or `@observable.prop` or `@mobx.decorators.prop` or ...

> we should preferably have one obvious way to do one thing. Not 5 ways that do almost the same thing.

I am not sure what you mean, because the words *"one thing"* and *"almost the same"* simply don't go together.

Either the things are exactly the same, like `@deep` and `@prop(deep())`, than you're right and my only argument taken from python zen is: *Although practicality beats purity.*
We could have only one way: `@prop(deep())` or `@observable(deep())`, however due to the fact it's something you would use most of the time, it seems logical to provide shorter, more explicit and readable version. In the end the `@prop(deep())` should be mainly a demonstration of how the composability work, but the preffered usage should be using explicit `@deep`

Or things are even slightly different and than I will go with: *Explicit is better than implicit.*

**Comment by antitoxic** [comment]
@urugator ok, so you don't agree. That's all I can contribute to the discussion. If you decide to go this way I will just wrap the module and abstract the interface. I currently never directly use external libraries. I wrap them into other objects which implement certain interfaces and write tests so that the library can be switched with another and its expected functionality can be verified. 

Namespacing a prop is a solution especially if I wrap mobx so it reads "observable.prop` . Though this wouldn't be my first choice. As I said that's not critical, I can wrap it.

Just to clarify "Making thing observable" is the one thing that I refer to when I say python's zen "do one thing in one way". What kind of observable is just a configuration. Just like when you read a file you want to have `open('file', mode=read)` and not `openForWriting(file)` ,`openForAppending(file)`, `openForRead(file)`

**Comment by urugator** [comment]
> If you decide to go this way

I won't be deciding anything, I am just trying to justify some ideas and confront them with others.

> wrapping

Won't you still need to wrap the modifiers themselves?

> "Making thing observable" is the one thing
 
"Making things observable" corelates with "Using Mobx" ... mobx is that single entry point, that one way, for "making things observable". 
But the need for modifiers proves that "making things observable" is a more complex problem and can't be done in one universal way.
So I just suggest to make this complexity transparent, by making things as explicit as possible.
Having a single all mighty magical "observable" function, doesn't help anyone in any way. You can't simply say "it makes things observable", because things can be observable in multiple ways. It just hides facts one still have to know. 

>  just a configuration

Matter of taste, I prefer using specific things rather than configuring generic ones. I prefer toolbox (mobx) with knife, saw, screwdriver, rather than toolbox with swissknife (observable).
Btw I don't say that the toolbox (mobx), can't have a compartments inside like: "tools for making things observable", "utilities", etc. Actually the decorators should be imho placed in own namespace.

> you want to have open('file', mode=read)

No, I don't want. It's horrible API and it's designed like this due to the technical reasons.
Many languages/libraries uses some kind of abstraction in the form of FileWriters, FileReaders and so on.

**Comment by mweststrate** [comment]
Closing this issue as mobx@3.0.0-rc.1 is now available, which addresses this issue. Changelog: https://github.com/mobxjs/mobx/blob/mobx3/CHANGELOG.md#300. The changes in more details: https://github.com/mobxjs/mobx/pull/725/files. Feel free to re-open if questions arise!
